File: Java/JugadorEspeculador.java
package civitas;

public class JugadorEspeculador extends Jugador {

    private static int FactorEspeculador = 2;
    private int fianza;

    public JugadorEspeculador(Jugador jug, int _fianza) {
        super(jug);
        fianza = _fianza;

        for (int i = 0; i < super.getPropiedades().size(); i++) {
            super.getPropiedades().get(i).actualizaPropietarioPorConversion(this);
        }
    }

    @Override
    protected boolean debeSerEncarcelado() {
        boolean b = super.debeSerEncarcelado();

        if (b) {
            b = super.puedoGastar(fianza);
        }

        return b;
    }

    @Override
    int getCasasMax() {
        return FactorEspeculador * super.getCasasMax();
    }

    @Override
    int getHotelesMax() {
        return FactorEspeculador * super.getHotelesMax();
    }

    @Override
    boolean pagaImpuesto(float cantidad) {
        return super.pagaImpuesto(cantidad / 2);
    }

    @Override
    public String toString() {
        return "Jugador Especulador \n" + super.toString() + "\n";
    }

}

File: Ruby/jugador_especulador.rb
module Civitas
  class JugadorEspeculador < Jugador
    
    @@FACTORESPECULADOR = 2
    def self.otro(jug, fianza)
      super(jug)
      @fianza = fianza
      
      for i in range super.propiedades.size
        super.propiedades[i].actualizarPropietarioPorConversion(self)
      end
    end
    
    def debeSerEncarcelado()
      encarcelar = super
      
      encarcelar = super.puedoGastar(@fianza) if (encarcelar)
      
      return encarcelar
    end
    
    def pagaImpuesto(cantidad)
      return super(cantidad/2)
    end
    
    def puedoEdificarCasa(propiedad)
      return propiedad.numCasas < @@CasasMax * @@factorEspeculador && super.puedoGastar(propiedad.precioEdificar)
    end
    
    def puedoEdificarHotel(propiedad)
      return propiedad.numHoteles < @@HotelesMax * @@factorEspeculador && super.puedoGastar(3 * propiedad.precioEdificar) && propiedad.numCasas >= @@CasasPorHotel
    end
    
    def to_s
      "Jugador Especulador \n" + super.to_s + "\n"
    end
  end
end
